Instructor: MCQ questions (single answer): support attaching weights to options #2776

sukanta-27 opened the pull request:
Fixes #2776 

**Outline of Solution**

**Backend implementation:**
- [x] Modified the `FeedbackMcqQuestionDetails` file to attach weights to mcq options and `other` option
- [x] Added tests in `InstructorFeedbackQuestionAddActionTest`
- [x] Added tests in `InstructorFeedbackQuestionEditActionTest`
- [x] Added unit tests for the `weight` feature in `FeedbackMcqQuestionDetailsTest`

**Front-end Part:** (UI demo: https://github.com/TEAMMATES/teammates/pull/8860#issuecomment-395554070)
- [x] Modified mcq form to add weight checkbox and weight cells
- [x] Modified `questionMcq.js`, `instructorFeedbackEdit.js` files
- [x] Added UI tests and End-to-end tests

**Mcq Statistics part:** (UI demo: https://github.com/TEAMMATES/teammates/pull/8860#issuecomment-397113482)
- [x] Add average to Overall `Response Summary`.
- [x] Add `Per Recipient Statistics` for mcq weights
- [ ] Add UI tests for mcq weights statistics
- [ ] Update CSV

**Update documentation**
- [ ] Update `InstructorHelpPage`

bqnguyen94 commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,288 @@
+        assertEquals(2, mcqDetails.getMcqChoices().size());
+        weights = mcqDetails.getMcqWeights();
+        assertTrue(weights.isEmpty());
+        assertEquals(0, weights.size());
```
Why is `weights` checked for emptiness twice here?

bqnguyen94 commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,288 @@
+        assertEquals(2, mcqDetails.getMcqChoices().size());
+        weights = mcqDetails.getMcqWeights();
+        assertFalse(weights.isEmpty());
+        assertEquals(2, weights.size());
```
I think asserting `isEmpty` is redundant?

sukanta-27 replied:
I guess you are right. I was trying to be extra cautious I guess :laughing: I guess only checking the size is okay then.

sukanta-27 commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -491,6 +491,9 @@ private Const() {
                 " is not a valid option for the " + Const.FeedbackQuestionTypeNames.MCQ + ".";
+        public static final String MCQ_ERROR_INVALID_WEIGHT =
+                "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.MCQ
+                + " must be valid numbers with precision up to 2 decimal places.";
```
I copied this error from `RUBRIC_ERROR_INVALID_WEIGHT`, so maybe for consistency keeping `must be valid numbers with precision up to 2 decimal places.` would be better? 

But I am good with whatever is decided in this matter.

bqnguyen94 replied:
I see. The current error message already can serve its purpose, so for consistency's sake we can keep it 👍

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -95,19 +116,75 @@ public boolean extractQuestionDetails(
+            String weight = HttpRequestHelper.getValueFromParamMap(
+                    requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHT + "-" + i);
+
+            if (choice != null && !choice.trim().isEmpty() && weight != null) {
```
In case if the user enters all spaces `"     "`.
The `extractQuestionDetails` method eliminates such choices from the list.
```java
                if (mcqChoice != null && !mcqChoice.trim().isEmpty()) {
                    mcqChoices.add(mcqChoice);
                    numOfMcqChoices++;
                }

```
 so I am also doing the same thing here. As, there is no point parsing the weight of a empty choice as it will anyway not pass validation.

xpdavid replied:
That's the validation logic rather than extraction logic I would say if we want clear definition between extractXXX and ValidateXXX.

sukanta-27 replied:
>That's the validation logic rather than extraction logic I would say if we want clear definition between extractXXX and ValidateXXX.

@xpdavid Sorry, I didn't understand this. :sweat_smile:

xpdavid replied:
Take a look at `extractQuestionDetails` and `validateQuestionDetails`.
I think It's OK to leave it there.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -95,19 +116,75 @@ public boolean extractQuestionDetails(
+    private double getMcqOtherWeight(Map<String, String[]> requestParameters,
+            boolean mcqOtherEnabled, boolean hasAssignedWeights) {
+
+        double mcqOtherWeight = this.mcqOtherWeight;
```
In case if the user tries to edit the question, and leaves the `otherWeight` field empty, then if I set `double mcqOtherWeight = 0` here, the existing otherWeight will be changed to zero, instead of preserving the previous weight.

xpdavid replied:
Yes. That's is the limitation of our update logic for question detail. However, I would say in the normal condition, we would always have the old `mcqOtherWeight` submitted in the UI.  Let's have this assumption until we have a better way to modify question details.

sukanta-27 replied:
Okay. Will use `0` here.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -457,11 +534,19 @@ public String getQuestionTypeChoiceOption() {
+            // If weights are enabled, number of choices and weights should be same.
+            if (hasAssignedWeights && mcqChoices.size() != mcqWeights.size()) {
+                errors.add(Const.FeedbackQuestion.MCQ_ERROR_INVALID_WEIGHT);
+            }
```
Oops! forgot about this case.

sukanta-27 commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,288 @@
+        assertEquals(0.0, mcqDetails.getMcqOtherWeight());
+        requestParams.clear();
+
+        ______TS("Success: weights enabled, other enabled, custom other weight");
```
@xpdavid Yes, I guess it should be.I forgot this case. I should check for `mcqWeights.size()` in `getMcqOtherWeight()` method? Till now, I was treating weight and otherWeight separately, but it should not be the case. :man_facepalming:

sukanta-27 commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,288 @@
+        assertTrue(mcqDetails instanceof FeedbackMcqQuestionDetails);
+        assertFalse(mcqDetails.hasAssignedWeights());
+        assertTrue(mcqDetails.getMcqWeights().isEmpty());
+        assertEquals(0.0, mcqDetails.getMcqOtherWeight());
```
Oh! Should I create different methods for each test case instead of doing multiple tests in a single `testAddWeights()` method then?

sukanta-27 commented:
@bqnguyen94 @xpdavid 
In attempt to resolve this problem https://github.com/TEAMMATES/teammates/pull/8860#discussion_r192562232, I added an assumption in `getMcqOtherWeight()` method:
`Assumption.assertNotNull("Null 'other' weight of MCQ question", weightOther);`

To test this, I added the following action test:

```java
    ______TS("Failed to edit 'other' option weight when new weight is null: Exception thrown");
// There is already responses for this question
assertFalse(frDb.getFeedbackResponsesForQuestion(fq.getId()).isEmpty());

String[] editNullOtherWeightsParams = {
        Const.ParamsNames.FEEDBACK_QUESTION_TEXT, "What do you like best about our product?",
        Const.ParamsNames.FEEDBACK_QUESTION_NUMBEROFCHOICECREATED, Integer.toString(mcqDetails.getNumOfMcqChoices()),
        Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHTS_ASSIGNED, "on",
        Const.ParamsNames.FEEDBACK_QUESTION_MCQCHOICE + "-0", mcqDetails.getMcqChoices().get(0),
        Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHT + "-0", "1",
        Const.ParamsNames.FEEDBACK_QUESTION_MCQCHOICE + "-1", mcqDetails.getMcqChoices().get(1),
        Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHT + "-1", "2",
        Const.ParamsNames.FEEDBACK_QUESTION_MCQOTHEROPTIONFLAG, "on",
        // Removed to send null for otherWeight parameter
        //Const.ParamsNames.FEEDBACK_QUESTION_MCQ_OTHER_WEIGHT, Double.toString(mcqDetails.getMcqOtherWeight())
};
requestedParams = new ArrayList<>(Arrays.asList(requiredParams));
Collections.addAll(requestedParams, editNullOtherWeightsParams);
a = getAction(requestedParams.toArray(new String[0]));
try {
    r = getRedirectResult(a);            
} catch (AssertionError e) {
    assertEquals("Null 'other' weight of MCQ question", e.getMessage());
}
// All existing response should remain
assertFalse(frDb.getFeedbackResponsesForQuestion(fq.getId()).isEmpty());

// Old value of otherWeight should be preserved due to keep existing policy
editedQuestion = FeedbackQuestionsLogic.inst()
        .getFeedbackQuestion(fs.getFeedbackSessionName(), fs.getCourseId(), 4);
editedQuestionDetails = (FeedbackMcqQuestionDetails) editedQuestion.getQuestionDetails();
assertEquals(5.0, editedQuestionDetails.getMcqOtherWeight());

```

**My query is:** Is it a good idea to catch assertionError or is there a better way to test this error?

xpdavid commented:
@sukanta-27 Here you go :) You can see the benefit of writing unit test :)
https://www.mkyong.com/unittest/testng-tutorial-2-expected-exception-test/

sukanta-27 commented:
@xpdavid I changed the mcq `questionMetaData` to the new format in `FeedbackSessionQuestionTypeTest`. Should I change the metadata for all mcq questions in all json files as well?

xpdavid commented:
> Should I change the metadata for all mcq questions in all json files as well?
Yes or no. We should consider the situation that some data in LIVE doesn't have those fields. Unless u have a plan do to data migration for LIVE data (e.g. adding those fields), we shall have the legacy data in test to make sure we can deal with such situation

sukanta-27 commented:
>we shall have the legacy data in test to make sure we can deal with such situation

Okay, I will leave it for now.

sukanta-27 commented:
@xpdavid @bqnguyen94 @damithc 

The UI demo of MCQ weights feature. 

**1.**

![ezgif com-gif-maker 1](https://user-images.githubusercontent.com/10681755/41124273-0c973626-6abf-11e8-91f5-71895b7db611.gif)

**2.**

![ezgif com-gif-maker 2](https://user-images.githubusercontent.com/10681755/41124306-3041a458-6abf-11e8-8f1f-6d228266f4f7.gif)

sukanta-27 commented:
@xpdavid @bqnguyen94 @damithc 
I have deployed the weight feature to my staging server: https://teammates-sukanta.appspot.com
Please have a look, play around with the mcq questions and provide feedback.

bqnguyen94 commented:
I'm good with this design, and the UI works nicely 👍

xpdavid commented:
Good work! @sukanta-27 
On a separate note. I forgot to mention this while reviewing backend part. Please also update `shouldChangesRequireResponseDeletion` in `questionDetails`. If feedbacks are already collected and instructor edit MCQ (e.g. adding weights). In this case, existing responses should be deleted

sukanta-27 commented:
@xpdavid 
>On a separate note. I forgot to mention this while reviewing backend part. Please also update shouldChangesRequireResponseDeletion in questionDetails. If feedbacks are already collected and instructor edit MCQ (e.g. adding weights). In this case, existing responses should be deleted

But as Prof. @damithc suggested in our discussions in Slack, we should be able to add weights even if there are existing responses in that particular question (like we can do in case of rubric weights), right?

xpdavid commented:
oh yeah. make sense👍

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +292,177 @@ public void testDeleteQuestionAction() {
+        assertNotNull(BackDoor.getFeedbackQuestion(courseId, feedbackSessionName, 1));
+
+        // verify html page
+        feedbackEditPage.verifyHtmlMainContent("/instructorFeedbackMcqQuestionWeightAddSuccess.html");
```
The above test case can be extracted to another method `testMcqWeightsFeature_shouldHaveCorrectDefaultValue`.

xpdavid replied:
You can separate your test also.
`testMcqWeightsFeature_shouldValidateFieldCorrectly`
`testMcqWeightsFeature_shouldAddQuestionCorrectly` - one failure one success should be enough
`testMcqWeightsFeature_shouldEditQuestionCorrectly` - one failure one success should be enough
`testMcqWeightsFeature_shouldToggleStateCorrectly`

sukanta-27 replied:
@xpdavid Actually, by this I thought you meant that I should extract my existing tests to separate methods and run them all together by running a single `@Test` named `testMcqWeightFeature`. So I did that. But as expected this way the tests are all still interdependent.

so, just to clarify Did you mean that only, or by separating the tests you meant separating them completely and putting them in separate `@Test` functions so that they are not dependent on one another?

xpdavid replied:
> putting them in separate @Test functions so that they are not dependent on one another?
If you can do that, it would be great.

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +292,177 @@ public void testDeleteQuestionAction() {
+        feedbackEditPage.clickSaveExistingQuestionButton(1);
+        feedbackEditPage.waitForTextsForAllStatusMessagesToUserEquals(Const.StatusMessages.FEEDBACK_QUESTION_EDITED);
+
+        ______TS("MCQ: Test weight is not added for invalid choice value");
```
The case is a bit "redundant" as u already has the same scenario in the integration/unit test. You can just have one failure case and that should be enough.

sukanta-27 replied:
>You can just have one failure case and that should be enough.

@xpdavid By this, do you mean that the failure tests I added in unit test is enough and I should delete this test? Or, do you mean, I should have only one failure case in <em>this</em> test only. (That is, to keep either the `"                "` (space only) choice or the empty choice case)?

xpdavid replied:
> do you mean that the failure tests I added in unit test is enough
Yes. Validate all scenarios in the unit tests should be enough.

You don't have to delete them I think as you have put lots of efforts in coming out and writing them.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,357 @@
+
+    @Test
+    public void testGetMcqWeights_allChoicesNull_weightsListShouldBeEmpty() {
+        ______TS("Failure: Weights enabled, null choices, custom weights, other disabled");
```
No need `______TS ` as your method name is already telling the story.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,357 @@
+    }
+
+    @Test
+    public void testValidateQuestionDetails_choicesLessThanMinRequirement_exceptionThrown() {
```
not `exceptionThrown` but errors returned right?

sukanta-27 replied:
Yes, I will change that to `errorsReturned`

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,357 @@
+        assertTrue(mcqDetails.hasAssignedWeights());
+        List<String> errors = mcqDetails.validateQuestionDetails(dummySessionToken);
+        assertEquals(0, errors.size());
+    }
```
Don't forget to cover all if conditions in the `validateQuestionDetails`. You can see the coverage by the CI or in your IDE.

sukanta-27 replied:
Most of the conditions are covered in tests, some of the branches which are missed are (I guess) not possible to test, as many of them are covered in methods `getMcqWeights` and `getMcqOtherWeight`. Like

```java
            if (!hasAssignedWeights && (!mcqWeights.isEmpty() || mcqOtherWeight != 0)) {
                errors.add(Const.FeedbackQuestion.MCQ_ERROR_INVALID_WEIGHT);
```
in this if statement, 2 branches out of 6 are missed, but if I set `hasAssignedWeights == false`, then `mcqWeights` will always be empty and `mcqOtherWeight` will always be 0 as that has been checked in the above two methods in the beginning. So, in this cases validation errors will not occur.

sukanta-27 replied:
Similarly, in the following if block:
```java
            if (hasAssignedWeights && !otherEnabled && mcqOtherWeight != 0) {
                errors.add(Const.FeedbackQuestion.MCQ_ERROR_INVALID_WEIGHT);
            }
```
One branch out of 6 is missed, but if any of `hasAssignedWeights` and `otherEnabled` are false, then `mcqOtherWeight` will always be zero, as that is covered in `getMcqOtherWeight` method.

sukanta-27 replied:
To be noted that, both `getMcqWeights()` and `getMcqOtherWeight()` are completely covered in the unit tests.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -95,19 +116,85 @@ public boolean extractQuestionDetails(
+
+        // If no other option has weights attached,
+        // the other option should not have a weight attached either
+        if (mcqWeights.isEmpty()) {
```
Okay.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -457,11 +569,34 @@ public String getQuestionTypeChoiceOption() {
-                && numOfMcqChoices < Const.FeedbackQuestion.MCQ_MIN_NUM_OF_CHOICES) {
-            errors.add(Const.FeedbackQuestion.MCQ_ERROR_NOT_ENOUGH_CHOICES
-                       + Const.FeedbackQuestion.MCQ_MIN_NUM_OF_CHOICES + ".");
+        if (generateOptionsFor == FeedbackParticipantType.NONE) {
```
Good point. :+1: Will do.

sukanta-27 replied:
@xpdavid I just realised that, validating this is redundant, as this situation will never occur. We can add the validation but we can't test it even if we pass `hasAssignedWeights` as `on` when `generateOptionsFor != NONE`, as the processing of weight fields and `hasAssignedWeights` are in the `generatedMcqOptions.equals(FeedbackParticipantType.NONE.toString()` block in `extractQuestionDetails` function. What should I do in this case.

```java
        if (generatedMcqOptions.equals(FeedbackParticipantType.NONE.toString())) {
            String numMcqChoicesCreatedString =
                    HttpRequestHelper.getValueFromParamMap(requestParameters,
                                                           Const.ParamsNames.FEEDBACK_QUESTION_NUMBEROFCHOICECREATED);
            Assumption.assertNotNull("Null number of choice for MCQ", numMcqChoicesCreatedString);
            int numMcqChoicesCreated = Integer.parseInt(numMcqChoicesCreatedString);

            for (int i = 0; i < numMcqChoicesCreated; i++) {
                String paramName = Const.ParamsNames.FEEDBACK_QUESTION_MCQCHOICE + "-" + i;
                String mcqChoice = HttpRequestHelper.getValueFromParamMap(requestParameters, paramName);
                if (mcqChoice != null && !mcqChoice.trim().isEmpty()) {
                    mcqChoices.add(mcqChoice);
                    numOfMcqChoices++;
                }
            }

            String hasAssignedWeightsString = HttpRequestHelper.getValueFromParamMap(
                    requestParameters, Const.ParamsNames.FEEDBACK_QUESTION_MCQ_HAS_WEIGHTS_ASSIGNED);
            boolean hasAssignedWeights = "on".equals(hasAssignedWeightsString);
            List<Double> mcqWeights = getMcqWeights(
                    requestParameters, numMcqChoicesCreated, hasAssignedWeights);
            double mcqOtherWeight = getMcqOtherWeight(requestParameters, mcqOtherEnabled, hasAssignedWeights);
            setMcqQuestionDetails(
                    numOfMcqChoices, mcqChoices, mcqOtherEnabled, hasAssignedWeights, mcqWeights, mcqOtherWeight);
        } else {
            setMcqQuestionDetails(FeedbackParticipantType.valueOf(generatedMcqOptions));
        }
```

sukanta-27 replied:
Instead of doing a validation for this, I wrote a unit test for `extractQuestionDetails` to test that the above mentioned case is covered in our logic.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -457,11 +569,34 @@ public String getQuestionTypeChoiceOption() {
-                && numOfMcqChoices < Const.FeedbackQuestion.MCQ_MIN_NUM_OF_CHOICES) {
-            errors.add(Const.FeedbackQuestion.MCQ_ERROR_NOT_ENOUGH_CHOICES
-                       + Const.FeedbackQuestion.MCQ_MIN_NUM_OF_CHOICES + ".");
+        if (generateOptionsFor == FeedbackParticipantType.NONE) {
```
Yes, negative weights are allowed.

sukanta-27 commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -37,13 +56,14 @@ function addMcqOption(questionNum) {
     const questionId = `#form_editquestion-${questionNum}`;
 
-    const $thisRow = $(`#mcqOptionRow-${index}-${questionNum}`);
+    const $thisRow = $(`.mcqChoiceRow-${index}-${questionNum}`);
```
I have used class so that I have a way to manipulate a specific choice and it's corresponding weight cell together. Each choice and it's corresponding weight cell falls under the same class.

bqnguyen94 replied:
You can also put an `id` on the enclosing `row` of the option and weight fragments.

bqnguyen94 replied:
Personally I prefer using `id` to get the choice and weight here since this class name is kind of inconsistent with the rest of the code, where `index` and `questionNum` are used in `id`s only.

xpdavid replied:
> where index and questionNum are used in ids only.
@sukanta-27 Yes. I think HTML class is not designed to be used like that

sukanta-27 replied:
@xpdavid @bqnguyen94 Okay. will use ids here.

sukanta-27 commented on src/test/java/teammates/test/cases/action/InstructorFeedbackQuestionEditActionTest.java:
```diff
@@ -580,6 +580,173 @@ public void testExecuteAndPostProcessMcq() {
+                        "FSQTT.idOfInstructor1OfCourse1",
+                        false),
+                r.getDestinationWithParams());
+        assertEquals(Const.StatusMessages.FEEDBACK_QUESTION_EDITED, r.getStatusMessage());
```
Yes we can add it. :+1:

bqnguyen94 commented on src/main/resources/feedbackQuestionMcqEditFormOptionFragment.html:
```diff
@@ -1,19 +1,17 @@
-      </button>
-    </span>
-  </div>
+<div class="input-group margin-bottom-7px mcqChoiceRow-${i}-${questionNumber}" id="mcqOptionRow-${i}-${questionNumber}">
```
Does the `id` here serve any purpose anymore?

sukanta-27 replied:
Yes, As I am now using `id` instead of class to remove options and weights in `removeMcqOptions`, so I am using this ID to get to the option row. I have removed the class though.

bqnguyen94 commented on src/main/resources/feedbackQuestionMcqEditFormTemplate.html:
```diff
@@ -1,8 +1,24 @@
+      <input type="checkbox" class="nonDestructive" id="${Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHTS_ASSIGNED}-${questionNumber}" name="mcqAssignWeights" ${mcqCheckAssignWeights}>
+      <span data-toggle="tooltip" data-placement="top" data-original-title="${Const.Tooltips.FEEDBACK_QUESTION_MCQ_ASSIGN_WEIGHTS}"> Choices are weighted </span>
+    </div>
+    <div class="col-sm-2 padding-left-45px" style="text-align:center;">
```
Can you put this style in the stylesheet?

bqnguyen94 commented on src/main/resources/feedbackQuestionMcqEditFormWeightFragment.html:
```diff
@@ -0,0 +1,5 @@
+<div class="margin-bottom-7px mcqChoiceRow-${i}-${questionNumber}">
```
If you are really going to use this class name in the end then rmb to put it in `Const`.

bqnguyen94 commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -25,7 +28,23 @@ function addMcqOption(questionNum) {
+        $(`#mcqWeights-${questionNum}`).html(weight);
+    } else {
+        $(choice).insertAfter(lastChoice);
+        $(weight).insertAfter(lastWeight);
```
How about assigning `$(`#mcqChoices-${questionNum}`)` to a variable and use `append` here?

sukanta-27 replied:
Done :+1:

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -338,7 +443,13 @@ public String getQuestionSpecificEditFormHtml(int questionNumber) {
+                Slots.MCQ_TOOLTIPS_ASSIGN_WEIGHT, Const.Tooltips.FEEDBACK_QUESTION_MCQ_ASSIGN_WEIGHTS,
+                Slots.MCQ_PARAM_ASSIGN_WEIGHT, Const.ParamsNames.FEEDBACK_QUESTION_MCQ_WEIGHTS_ASSIGNED,
+                Slots.MCQ_EDIT_FORM_WEIGHT_FRAGMENTS, weightFragmentHtml.toString(),
+                Slots.MCQ_PARAM_OTHER_WEIGHT, Const.ParamsNames.FEEDBACK_QUESTION_MCQ_OTHER_WEIGHT,
```
@xpdavid But should I add a `HAS` here given that it is not boolean? It is the ID of the other weight cell. That is `mcqOtherWeight`

xpdavid replied:
Then no need :D

sukanta-27 commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.js:
```diff
@@ -355,15 +359,21 @@ function disableQuestion(questionNum) {
     /* Check whether generate options for students/instructors/teams is selected
-       If so, hide 'add Other option' */
+       If so, hide 'add Other option' and the mcq weight checkbox for mcq questions */
     if ($currentQuestionTable.find(`#generateMcqOptionsCheckbox-${questionNum}`).prop('checked')) {
```
`toggleMcqGeneratedOptions` has many operations in it, we won't need all of it in `disableQuestion`, (specifically the choice table). Using `toggleMcqGeneratedOptions` will cause following problem for MCQ and MSQ questions.

![screenshot 507](https://user-images.githubusercontent.com/10681755/41233425-085d0226-6da7-11e8-9e89-76f8427af4d4.png)
![screenshot 508](https://user-images.githubusercontent.com/10681755/41233427-08ab1f24-6da7-11e8-849b-8d3171add263.png)

sukanta-27 commented:
@xpdavid @bqnguyen94 Changes have been made. Ready for review.

bqnguyen94 commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -7,9 +7,12 @@ function addMcqOption(questionNum) {
-    $(`
-    <div class="margin-bottom-7px" id="mcqOptionRow-${curNumberOfChoiceCreated}-${questionNum}">
+    const choiceFragment = (`
+    <div class="margin-bottom-7px mcqChoiceRow-${curNumberOfChoiceCreated}-${questionNum}"
```
You forgot this class name and in the weight fragment below :P

sukanta-27 replied:
Oops! Fixing it.

sukanta-27 commented:
@xpdavid @bqnguyen94 @damithc The MCQ statistics UI:

**For cases where weights are attached** 
![screenshot 514](https://user-images.githubusercontent.com/10681755/41382561-44600112-6f8a-11e8-9a27-9b1f75dd2e20.png)

**Note:** In the `Per Recipient Statistics`, I have added the weights in this format `Choice [Weight]`, this is to save some space by not writing `Choice (Weight: x)`. Also, as the `Response Summary` already shows corresponding weights of each option, I thought adding `(Weights: x)` again in the table below will be redundant. Also, I haven't done it yet, but I will be sorting the `Per Recipient Statistics` table as we have in rubric questions.

**For cases where weights are not attached**
![screenshot 515](https://user-images.githubusercontent.com/10681755/41382570-4ed2f46a-6f8a-11e8-894a-c974a70f6e0f.png)


**Note:** The `Per Recipient statistics` is not shown if weights are not attached.

sukanta-27 commented:
**Sorted Per Response Statistics Tabled UI:**

![screenshot 520](https://user-images.githubusercontent.com/10681755/41440686-3004ca2c-704d-11e8-8f0d-c17b96ad0874.png)

damithc commented:
> Sorted Per Response Statistics Tabled UI:

Can the columns be sortable?

sukanta-27 commented:
>Can the columns be sortable?

I can try that.

sukanta-27 commented:
Updated CSV stats for MCQ questions. 

**For Questions without weights attached, the csv is the same as before**
![screenshot 527](https://user-images.githubusercontent.com/10681755/41511569-ede7be14-7296-11e8-95b2-3f9a507a07a1.png)


**For Questions with weights attached:**
![screenshot 526](https://user-images.githubusercontent.com/10681755/41511572-ff6c4d6c-7296-11e8-8ba8-286afca37f8a.png)

sukanta-27 commented:
@xpdavid @bqnguyen94 Ready for review.

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+        feedbackEditPage.fillMcqWeightBox(1, 3, "5");
+
+        feedbackEditPage.clickSaveExistingQuestionButton(1);
+        feedbackEditPage.waitForTextsForAllStatusMessagesToUserEquals(Const.StatusMessages.FEEDBACK_QUESTION_EDITED);
```
No error message?

sukanta-27 replied:
No. This is the normal behaviour for cases where there are invalid (space-only) choices. The question form just simply removes the choice cells with the invalid value. SO, this is also the default behaviour for the corresponding weight cell as well.

This is the reason, I am testing that the choice cell and the corresponding weight cell are not present in the question form after editing the question.

```java
        // Check that the two added choices and the corresponding weights are not present
        assertFalse(feedbackEditPage.isElementPresent("mcqOption-3-1"));
        assertFalse(feedbackEditPage.isElementPresent("mcqWeight-3-1"));
```

xpdavid replied:
Add comment to note it down

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+        // Check that weight fields are again visible after unchecking generated option,
+        // and also check that all the fields have the default values stored.
+        assertFalse(feedbackEditPage.isMcqAssignWeightCheckboxChecked(1));
+        assertFalse(browser.driver.findElement(By.id("mcqOtherOptionFlag-1")).isSelected());
```
Find existing method or create a new method to check is selected?

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+        // Add One option and check MCQ assign weight checkbox to check the mcq weight field value.
+        feedbackEditPage.clickAddMoreMcqOptionLink(1);
+        feedbackEditPage.clickMcqAssignWeightsCheckbox(1);
+        assertTrue(feedbackEditPage.isElementVisible("mcqWeight-0-1"));
```
Same here? There are other places in your PR.

xpdavid commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+    }
+
+    @Test(priority = 1)
+    public void testMcqWeightsFeature_instructorResultsPageQuestionView_showStatistics() throws Exception {
```
I believe this test should be `InstructorFeedbackResultsPageUiTest` here?

sukanta-27 replied:
I also thought that initially, but decided against it when I saw Rubric questions also have the results page tests in `FeedbackRubricQuestionUiTest` file, and one other reason was that I couldn't find any specific question related test in `InstructorFeedbackResultsPageUiTest`, all the tests seems more generic instead of focusing on a single aspect of a single question.

So should I move this test in `InstructorFeedbackResultsPageUiTest`? Wouldn't it be better if specific question related tests belong to a particular file?

xpdavid replied:
Sure, cat put it here. Remember to follow the convention in your later PRs.

xpdavid commented on src/test/java/teammates/test/cases/datatransfer/FeedbackMcqQuestionDetailsTest.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    @Test
+    public void testValidateQuestionDetails_weightsAndOtherEnabledEmptyWeights_errorReturned() {
```
The execution path will hit `choices.size() != weight.size()` right?

sukanta-27 replied:
@xpdavid Oops! You are right! This hits the above condition, along with the case I wrote this test for.
```
            if (hasAssignedWeights && !otherEnabled && mcqOtherWeight != 0) {
                errors.add(Const.FeedbackQuestion.MCQ_ERROR_INVALID_WEIGHT);
            }
```
As I said in an earlier review comment that this case seems impossible in the end product, I should remove this case right?

xpdavid replied:
Sure, cat put it here. Remember to follow the convention in your later PRs.

xpdavid commented on src/test/resources/data/FeedbackMcqQuestionUiTest.json:
```diff
@@ -55,6 +69,38 @@
+      "team": "Team 1"
+    },
+    "student3.tmms@FMcqQnUiT.CS2104": {
+      "googleId": "FMcqQnUiT.student3",
```
No such googID?

sukanta-27 replied:
Actually, I just checked the other student's data (Alice) in this file and gave a similar googleID to these students.

```json
   "alice.tmms@FMcqQnUiT.CS2104": {
       "googleId": "FMcqQnUiT.alice.tmms",
       "email": "alice.b.tmms@gmail.tmt",
       "course": "CFeedbackEditUiT.CS2104",
       "name": "Alice Betsy</td></div>'\"",
       "comments": "This student's name is Alice Betsy</td></div>'\"",
       "team": "Team </option><option value=\"dump\">1'"
    },
```
I thought we can put any value here :sweat_smile: . Am I missing something?

xpdavid replied:
Alice has `account` entity

xpdavid replied:
A student can have no google ID (Not yet join / access through token)

https://github.com/TEAMMATES/teammates/blob/master/src/main/java/teammates/storage/entity/CourseStudent.java#L51

But if there is a googleID, there must be an account.

xpdavid commented on src/test/resources/data/FeedbackMcqQuestionUiTest.json:
```diff
@@ -55,6 +69,38 @@
+      "team": "Team 1"
+    },
+    "student4.tmms@FMcqQnUiT.CS2104": {
+      "googleId": "FMcqQnUiT.student4",
```
Same here?

xpdavid commented on src/test/resources/data/FeedbackSessionQuestionTypeTest.json:
```diff
@@ -631,7 +631,7 @@
       "creatorEmail": "instructor1@course1.tmt",
       "questionMetaData": {
-        "value": "{\"numOfMcqChoices\":2,\"mcqChoices\":[\"It's good\",\"It's perfect\"],\"questionText\":\"What do you like best about the class' product?\",\"questionType\":\"MCQ\",\"otherEnabled\":false}"
+        "value": "{\"hasAssignedWeights\": false,\"mcqWeights\": [],\"mcqOtherWeight\": 0.0,\"numOfMcqChoices\": 2,\"mcqChoices\": [\"It's good\",\"It's perfect\"],\"otherEnabled\": false,\"generateOptionsFor\": \"NONE\",\"questionType\": \"MCQ\",\"questionText\": \"What do you like best about the class' product?\"}"
```
Have you deliberately left some metadata without `mcqWeights` etc?

sukanta-27 replied:
Yes, instead of modifying the existing questions to add weights, I created a new question for that purpose, in this way, all the cases which used the previous questions will not get affected,

sukanta-27 commented:
@xpdavid Thank you very much. :smile:  I haven't deployed everything on my staging server yet. Let me update you when I deploy it.

bqnguyen94 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
-                            </span>
-                              </div>
-                            </div>
+              <div class="col-sm-12 margin-bottom-7px">
```
Is something missing here? Else this should be indented differently

bqnguyen94 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
+                  <div class="row">
+                    <div class="col-sm-9" id="mcqChoices-1">
+                      <div class="input-group margin-bottom-7px" id="mcqOptionRow-0-1">
+              <span class="input-group-addon">
```
And here

bqnguyen94 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
+                    <div class="col-sm-9" id="mcqChoices-1">
+                      <div class="input-group margin-bottom-7px" id="mcqOptionRow-0-1">
+              <span class="input-group-addon">
+                <input type="radio" class="disabled_radio" disabled="">
```
Can you change the class to dash-cased?

sukanta-27 replied:
@bqnguyen94 Do you mean like `disabled-radio`? But the original class name is `disabled_radio` only? Don't you think not having the original class name might create confusion afterwards?

bqnguyen94 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
           <p>
             When you view the results of a multiple-choice (single answer) question, TEAMMATES calculates some statistics about the results collected,
             such as the number of responses for each option, and the percentage of response in which each option was chosen.
+            If weights are assigned in the question, then additional statistics are calculated. Such as, average point for each option.
```
`... additional statistics are calculated such as count and average point for each option.`

bqnguyen94 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
+          </p>
+          <p>
+            <strong>Note: </strong> Weights can be assigned for questions with existing questions too. To do that, edit the question that needs weights,
+            and if there are existing responses for that questions, the additional statistics will automatically get calculated for the question.
```
`... if there are existing responses for that question ...`

sukanta-27 commented:
@bqnguyen94 @xpdavid Instructor Help page: https://teammates-sukanta.appspot.com/instructorHelp.jsp#questions

xpdavid commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
+                    </form>
+                    <div class="inline panel-heading-text">
+                      <!--Note: When an element has class text-preserve-space, do not insert and HTML spaces-->
+                      <span class="text-preserve-space">Did you understand today's lecture?&nbsp;<span style=" white-space: normal;">
```
I don't think this is a good question to demo the weight feature.

It doesn't make sense to talk about per recipient statistics for self-feedback question right?

sukanta-27 replied:
You are right. I should have noticed it. Changing it now.

sukanta-27 commented:
> perhaps you can try using a text editor that have auto-beautifying function?

@xpdavid @bqnguyen94 The `instructorHelpQuestions` page has a lot of places where there are wrong indentations. Can I reindent the whole file? Will that be acceptable? Or, should I just reindent the changes I have made?

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -397,28 +503,185 @@ public String getQuestionAdditionalInfoHtml(int questionNumber, String additiona
 
     @Override
-    public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> responses,
+    public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> unsortedResponses,
```
@xpdavid Like we do in Rubric question? Okay. I will do that.

bqnguyen94 commented:
> Can I reindent the whole file?

Yes I think this is ideal, making it easier for the next guy to work on the file.

sukanta-27 commented:
@xpdavid @bqnguyen94 I have modified the example question. Please have a look again at my [staging server](https://teammates-sukanta.appspot.com/instructorHelp.jsp#questions)

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -516,4 +923,89 @@ public String validateGiverRecipientVisibility(FeedbackQuestionAttributes feedba
+     * Calculates the average weight of each option based on the all responses.
+     * Weights should be assigned or else an empty map will be returned.<br>
+     * The average of an option is calculated as follows:<br><br>
+     * Average = (frequency of the option based on responses) * (weight of the option) / total responses.
```
@xpdavid To be honest, I am not sure what is considered the correct way of using the weight. But this portion of code (Calculating the average for each option based on their weight and response count) is similar to what we show in rubric question stats. 
![screenshot 528](https://user-images.githubusercontent.com/10681755/41614515-53883b7e-7416-11e8-9c8d-bb78ecec402a.png)

xpdavid replied:
@damithc Is this the correct way of calculating weights?

damithc replied:
@sukanta-27 can give a specific example scenario (question, options, responses) and the stats calculated? That way we can get a sense of whether the stats make sense.

sukanta-27 replied:
@damithc @xpdavid Here is an example for MCQ question

**Question:**
![screenshot 529](https://user-images.githubusercontent.com/10681755/41645564-e4f7d688-7460-11e8-882e-5667d20321b0.png)

**Responses:**
![screenshot 531](https://user-images.githubusercontent.com/10681755/41645628-0929f036-7461-11e8-8b75-3c7a93fb41ae.png)

**Stats:**
![screenshot 530](https://user-images.githubusercontent.com/10681755/41645669-277b0aca-7461-11e8-9449-7f1b0362a4ab.png)

damithc replied:
@sukanta-27 thanks. the doubt is regarding the `Response Summary` section and not the `Per Recipient Statistics`? Which columns?

sukanta-27 replied:
>the doubt is regarding the Response Summary section and not the Per Recipient Statistics?

@damithc Yes, I guess the doubt is regarding the `Response Summary` section, Average coulmn. @xpdavid Can you confirm?

xpdavid replied:
Yes. Actually, I am curious why the denominator doesn't have the `totalWeight` factor.

sukanta-27 replied:
@xpdavid You mean to say:
`Average = (response count per option) * (weight of that option) / total Weight`?

damithc replied:
It's hard to give a meaning to the `Average` column. I suspect users will be equally confused about what it means. Perhaps omit `Weight` and `Average` columns altogether and show only `Percentage(count)`? i.e., weights are not considered in this table at all.

xpdavid replied:
Something like that. I think as long as the stats help instructors to make decisions, it should be good. 

Actually, weighted Percentage can also be a good option.

WeightedPercentage = (response count per option) * (weight of that option) / [(response count for option i) * (weight of option i) for i to n]

Let's say if no weight (same as weight = 1), this will be the same as a percentage. If some option is not helpful (like `Not Sure` in feedback), instructors can give it lower weight. We can also see the difference between normal percentage and weighted percentage.

sukanta-27 replied:
I think having two kind of percentage can confuse the instructor? Isn't that so? Specially when we don't show that kind of stat anywhere else.

damithc replied:
>Actually, weighted Percentage can also be a good option.

Yes, that can be useful.

damithc replied:
> I think having two kind of percentage can confuse the instructor?

We'll have to clearly indicate which column is weighted and which is not. Actually, it's best to show both or none at all; otherwise the user will wonder which one it is, misinterpret one as the other, or not even realize there are two types of percentages possible.

sukanta-27 replied:
@xpdavid @damithc Okay, in that case, I will replace the `Average` column with `Weighted percentage` column. Will also add some details about the weighted percentage in the instructor help page.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -457,10 +836,38 @@ public String getQuestionTypeChoiceOption() {
+
+            // If no other option has weights attached,
+            // the 'other' option should not have a weight attached either
+            if (hasAssignedWeights && otherEnabled && mcqWeights.isEmpty() && mcqOtherWeight != 0) {
```
>In what situation will we have this case?

I am not sure. This case seems impossible in the end product, as all the weights have front-end validation attached to them, I am not sure how we will get `mcqWeights.isEmpty()` situation. As, the form will not submit until all the weight cells have some value.

xpdavid replied:
Can delete it.

sukanta-27 commented on src/main/resources/feedbackQuestionMcqResultStatsTemplate.html:
```diff
@@ -7,25 +7,32 @@
-    <div class="col-sm-12">
-      <table class="table margin-0">
+    <div class="col-sm-12 table-responsive">
+      <table class="table table-bordered table-striped margin-0">
```
>Can this table

Yes, I can make them sortable.

>should this table be sortable?

IMHO I think even if we do that, it will not be used a lot. 
@xpdavid @bqnguyen94 @damithc What do you think? Should we do it?

xpdavid replied:
If there are many options and instructors want to sort them based on percentage or average. This might be helpful.

damithc replied:
Yes, we can make it sortable if it's not too much extra work.

sukanta-27 replied:
Okay. Will make them sortable.

sukanta-27 commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -37,54 +55,166 @@ function addMcqOption(questionNum) {
         $(`#mcqChoiceTable-${questionNum}`).hide();
+        $(`#mcqAssignWeights-${questionNum}`).parent().hide();
+        // Hide the 'Weights' label
+        $(`#mcqAssignWeights-${questionNum}`).parent().siblings('div').hide();
```
@xpdavid That is because, when we change from generate option to regular mcq choices, all the values like weights, choices get reset to default, and if I bring the label back in the else, then, the label will show when it shouldn't.

 And  this case is covered anyway, as in the else block, I am calling `toggleMcqAssignWeights($(`#mcqAssignWeights-${questionNum}`), questionNum);` which will automatically show the weight label if the checkbox is checked.

sukanta-27 commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -37,54 +55,166 @@ function addMcqOption(questionNum) {
     }
 }
 
+function bindMcqOtherOptionEnabled() {
```
Okay.

sukanta-27 commented on src/main/webapp/dev/js/main/instructorFeedbackEdit.js:
```diff
@@ -358,15 +362,21 @@ function disableQuestion(questionNum) {
+       If so, hide 'add Other option' and the mcq weight checkbox for mcq questions */
     if ($currentQuestionTable.find(`#generateMcqOptionsCheckbox-${questionNum}`).prop('checked')) {
         $currentQuestionTable.find(`#mcqOtherOptionFlag-${questionNum}`).closest('.checkbox').hide();
+        $currentQuestionTable.find(`#mcqAssignWeights-${questionNum}`).parent().hide();
```
`.parent()` is the enclosing div for both the checkbox and the `Choices are weighted` label. Okay, will add a comment here.

sukanta-27 commented on src/main/webapp/partials/instructorHelpQuestions.jsp:
```diff
@@ -358,589 +358,1614 @@
+              (Optional) Specify 'Other' option, and let the student enter their own answer
+            </li>
+            <li>
+              (Optional) Assign weights to each option for calculating statistics
```
I have mentioned it in the following paragraphs after the question edit form example.

sukanta-27 commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+        // Check validation for MCQ weight cells
+        feedbackEditPage.fillMcqWeightBox(NEW_QUESTION_INDEX, 0, "");
+        feedbackEditPage.clickAddQuestionButton();
+        assertTrue(feedbackEditPage.isMcqWeightBoxFocused(NEW_QUESTION_INDEX, 0));
```
>You should fill with a valid value for the corresponding MCQ options before checking this invalid case?

You mean first adding a question with valid weights and then editing them to provide the invalid weights?

>Actually, this is tricky when more and more required fields are added in the project base. Try to see if you can find a better way to validate the required field behavior.

Okay, will try to do that.

sukanta-27 commented on src/test/java/teammates/test/cases/browsertests/FeedbackMcqQuestionUiTest.java:
```diff
@@ -291,4 +294,294 @@ public void testDeleteQuestionAction() {
+        feedbackEditPage.verifyHtmlMainContent("/instructorFeedbackMcqQuestionWeightAddSuccess.html");
+
+        // Delete the question
+        feedbackEditPage.clickDeleteQuestionLink(1);
```
As all the weight tests are on separate `@Test`, It is hard to predict which tests run after which tests, so if I leave the question without deleting it, I will have to change the question numbering (for IDs or for asserts whatever) on every test case in the whole file and It may bring a lot of inconsistencies due to the unpredictableness of the execution order of tests. SO, I just delete each question after a test completes to not hamper the whole test suite, and so that we can easily add more tests later without thinking about the inter-dependencies of all the previous tests.

sukanta-27 commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -397,28 +503,185 @@ public String getQuestionAdditionalInfoHtml(int questionNumber, String additiona
+        }
+
+        // Calculate responses of each option and increment the value of perRecipientResponses value.
+        for (FeedbackResponseAttributes response : responses) {
```
@xpdavid I am not very good with lambda functions. :sweat_smile: 

>The first `computeIfAbsent` is to put recipient to `responseCountPerOption` mapping

Do you mean like this? 

```java
        responses.forEach((response) -> {
            Map<String, Integer> responseCount = new LinkedHashMap<>();
            for (String choice : mcqChoices) {
                responseCount.put(choice, 0);
            }
            if (otherEnabled) {
                responseCount.put("Other", 0);
            }

            perRecipientResponse.computeIfAbsent(response.recipient, responseCountPerOption -> responseCount);
        });

```

xpdavid replied:
```java
        responses.forEach((response) -> {
            perRecipientResponse.computeIfAbsent(response.recipient, key -> {
				// construct default value for responseCount
			});
			perRecipientResponse.computeIfPresent(response.recipient, (key, responseCountPerOption) -> {
				// update responseCount here
			});
        });
```

sukanta-27 commented:
Made the suggested changes. Ready for review.

xpdavid commented:
The staging server doesn't have update weighted percentage in instructor help page? @sukanta-27

sukanta-27 commented:
Oops! I always forget to update the staging server :sweat_smile:

sukanta-27 commented:
@xpdavid I have updated my staging server. Also, for testing the statistics part, I have changed the feedback path for `Test Session` to `Instructors -> Other instructor's in the course` and submitted my responses. You can moderate responses of others to test the result stats.

sukanta-27 commented:
>I am still wondering why we should allow negative weight?

I allowed negative weights for two reasons. One being it is also allowed in rubric weights so I thought it is what is expected for mcq weights as well. And my second reasoning was, I thought allowing negative weights may give instructors a little more flexibility, as if we consider weights as points added for a student, negative weights for an option can be considered as points deducted from the student.

>In addition, in the situation where all weights are zero, there will be a bug - https://teammates-sukanta.appspot.com/page/instructorFeedbackResultsPage?courseid=CS101&fsname=Test+session

Oops! Looking into it.

xpdavid commented:
> as if we consider weights as points added for a student, negative weights for an option can be considered as points deducted from the student.

- Weighted percentage won't make any sense here I believe.
- Instructors can always use the positive weight to achieve the same thing (as they care more about the relative average (compare to the peers) rather than the absolute average)

sukanta-27 commented:
@xpdavid Changes are made. Ready for review.

xpdavid commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -741,8 +743,8 @@ public String validateGiverRecipientVisibility(FeedbackQuestionAttributes feedba
                 double weight = weightedPercentagePerOption.get(key);
-                double weightedPercentage =
-                        100 * ((frequency * weight) / calculateTotalWeightedResponseCount(answerFrequency));
+                double weightedPercentage = weight == 0 ? 0
```
The check is redundant I think.

zero times a rational number is zero. zero divides rational number is also zero.

sukanta-27 replied:
The bug you mentioned in https://github.com/TEAMMATES/teammates/pull/8860#pullrequestreview-131410993 originates from here. When `totalWeightedResponseCount == 0`, this bug occurs. Though I can modify this check to test for  `totalWeightedResponseCount == 0` only. 

>zero divides rational number is also zero.

Umm. Isn't 0/0 `undefined`? That is why the bug is occuring. Though it is often confusing, I found this article which explains this nicely.
https://brilliant.org/wiki/what-is-0-0/

xpdavid replied:
Then your check should be done on `totalWeightedResponseCount` rather than `weight`

sukanta-27 replied:
Yes, that is what I am doing, I have changed the `weight` to `totalWeightedResponseCount`.

xpdavid commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -495,7 +495,7 @@ private Const() {
         public static final String MCQ_ERROR_INVALID_WEIGHT =
                 "The weights for the choices of a " + Const.FeedbackQuestionTypeNames.MCQ
-                + " must be valid numbers with precision up to 2 decimal places.";
+                + " must be valid positive numbers with precision up to 2 decimal places.";
```
must be valid non-negative numbers (since we allow zero to exist)

xpdavid commented on src/main/webapp/dev/js/common/questionMcq.js:
```diff
@@ -106,26 +121,21 @@ function toggleMcqHasAssignedWeights($checkbox, questionNum) {
-        if ($otherEnabledCheckbox.prop('checked')) {
-            $otherWeightCell.show();
-        }
+        toggleVisibilityOfMcqOtherWeight($otherEnabledCheckbox, questionNum);
```
If we do the same thing in `if` and `else`, we shall put it outside.

sukanta-27 commented:
Made the suggested changes. Ready for review.

damithc commented on src/main/java/teammates/common/datatransfer/questions/FeedbackMcqQuestionDetails.java:
```diff
@@ -457,10 +597,47 @@ public String getQuestionTypeChoiceOption() {
+            }
+
+            // If weights are enabled, and any of the weights have negative value, this error will
+            // be triggered.
```
A note for future: usually this type of comments are written in imperative mood. 
e.g., `// If weights are enabled, and any of the weights have negative value, trigger this error`

sukanta-27 commented:
@damithc Yes sir,  manual testing has been done.

sukanta-27 commented:
@xpdavid @bqnguyen94  I would suggest to give me a day before merging this, as right now I am away from my laptop. I will fix the comments and after that notify you guys?

sukanta-27 commented:
@xpdavid @bqnguyen94 Modified the comments.
